src/server/active_tran_server.cpp

```diff
@@ -351,23 +369,34 @@ void active_tran_server::receive_data_page (cubpacking::unpacker &upk)
   std::string message;
   upk.unpack_string (message);
 
-  int error_code;
-  std::memcpy (&error_code, message.c_str (), sizeof (error_code));
+  if (message.size () == sizeof (int))
+    {
+      // We have an error.
+      int error_code;
+      std::memcpy (&error_code, message.c_str (), sizeof (error_code));
 
-  if (prm_get_bool_value (PRM_ID_ER_LOG_READ_DATA_PAGE))
+      if (prm_get_bool_value (PRM_ID_ER_LOG_READ_DATA_PAGE))
+	{
+	  _er_log_debug (ARG_FILE_LINE, "Received data page message from Page Server. Error: %d \n", error_code);
+	}
+    }
+  else
     {
-      if (error_code == NO_ERROR)
+      // We have a page.
+      auto shared_data_page = std::make_shared<std::string> (message);
+
+      auto io_page = (FILEIO_PAGE *)shared_data_page->c_str ();
+      VPID id;
+      id.pageid = io_page->prv.pageid;
+      id.volid = io_page->prv.volid;
+
+      m_data_page_broker->set_page (id, std::move (shared_data_page));
+
+      if (prm_get_bool_value (PRM_ID_ER_LOG_READ_DATA_PAGE))
 	{
-	  char buf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
-	  FILEIO_PAGE *io_page = (FILEIO_PAGE *) PTR_ALIGN (buf, MAX_ALIGNMENT);
-	  std::memcpy (io_page, message.c_str () + sizeof (error_code), db_io_page_size ());
 	  _er_log_debug (ARG_FILE_LINE, "Received data page message from Page Server. LSA: %lld|%d, Page ID: %ld, Volid: %d",
 			 LSA_AS_ARGS (&io_page->prv.lsa), io_page->prv.pageid, io_page->prv.volid);
```

Comment:
Move the declaration of io_page to the scope where it is used.
```suggestion
	  auto io_page = reinterpret_cast<const FILEIO_PAGE *> (shared_data_page->c_str ());
	  _er_log_debug (ARG_FILE_LINE, "Received data page message from Page Server. LSA: %lld|%d, Page ID: %ld, Volid: %d",
			 LSA_AS_ARGS (&io_page->prv.lsa), io_page->prv.pageid, io_page->prv.volid);
```